p2p/transport/tcp/conn/connection.go

```diff
@@ -269,147 +233,181 @@ func (c *MConnection) stopServices() (alreadyStopped bool) {
 		<-c.doneSendRoutine
-
 		// Send and flush all pending msgs.
 		// Since sendRoutine has exited, we can call this
 		// safely
 		w := protoio.NewDelimitedWriter(c.bufConnWriter)
-		eof := c.sendSomePacketMsgs(w)
+		eof := c.sendBatchPacketMsgs(w, numBatchPacketMsgs)
 		for !eof {
-			eof = c.sendSomePacketMsgs(w)
+			eof = c.sendBatchPacketMsgs(w, numBatchPacketMsgs)
 		}
-		c.flush()
-
-		// Now we can close the connection
+		_ = c.flush()
 	}
 
-	c.conn.Close()
-
-	// We can't close pong safely here because
-	// recvRoutine may write to it after we've stopped.
-	// Though it doesn't need to get closed at all,
-	// we close it @ recvRoutine.
-
-	// c.Stop()
+	return c.conn.Close()
 }
 
-// OnStop implements BaseService.
-func (c *MConnection) OnStop() {
-	if c.stopServices() {
-		return
-	}
+func (c *MConnection) ConnState() (state transport.ConnState) {
+	state.ConnectedFor = time.Since(c.created)
+	state.SendRateLimiterDelay = c.sendMonitor.Status().SleepTime
+	state.RecvRateLimiterDelay = c.recvMonitor.Status().SleepTime
+	state.StreamStates = make(map[byte]transport.StreamState)
 
-	c.conn.Close()
+	for streamID, channel := range c.channelsIdx {
+		state.StreamStates[streamID] = transport.StreamState{
+			SendQueueSize:     channel.loadSendQueueSize(),
+			SendQueueCapacity: cap(channel.sendQueue),
+		}
+	}
 
-	// We can't close pong safely here because
-	// recvRoutine may write to it after we've stopped.
-	// Though it doesn't need to get closed at all,
-	// we close it @ recvRoutine.
+	return state
 }
 
 func (c *MConnection) String() string {
 	return fmt.Sprintf("MConn{%v}", c.conn.RemoteAddr())
 }
 
-func (c *MConnection) flush() {
-	c.Logger.Debug("Flush", "conn", c)
-	err := c.bufConnWriter.Flush()
-	if err != nil {
-		c.Logger.Debug("MConnection flush failed", "err", err)
-	}
+func (c *MConnection) flush() error {
+	return c.bufConnWriter.Flush()
 }
 
 // Catch panics, usually caused by remote disconnects.
 func (c *MConnection) _recover() {
 	if r := recover(); r != nil {
 		c.Logger.Error("MConnection panicked", "err", r, "stack", string(debug.Stack()))
-		c.stopForError(fmt.Errorf("recovered from panic: %v", r))
-	}
-}
-
-func (c *MConnection) stopForError(r any) {
-	if err := c.Stop(); err != nil {
-		c.Logger.Error("Error stopping connection", "err", err)
-	}
-	if atomic.CompareAndSwapUint32(&c.errored, 0, 1) {
-		if c.onError != nil {
-			c.onError(r)
-		}
+		c.Close(fmt.Sprintf("recovered from panic: %v", r))
 	}
 }
 
-// Queues a message to be sent to channel.
-func (c *MConnection) Send(chID byte, msgBytes []byte) bool {
+// thread-safe.
+func (c *MConnection) sendBytes(chID byte, msgBytes []byte, blocking bool) error {
 	if !c.IsRunning() {
-		return false
+		return nil
 	}
 
-	c.Logger.Debug("Send", "channel", chID, "conn", c, "msgBytes", log.NewLazySprintf("%X", msgBytes))
+	// c.Logger.Debug("Send",
+	// 	"streamID", chID,
+	// 	"msgBytes", log.NewLazySprintf("%X", msgBytes),
+	// 	"timeout", timeout)
```

Comment:
```suggestion
	// c.Logger.Debug("Send",
	// 	"streamID", chID,
	// 	"msgBytes", log.NewLazySprintf("%X", msgBytes),
	// 	"timeout", timeout)
```
Do we want to keep this debug message?

Reply:
It's only useful if you're a CometBFT dev debugging the P2P stack. Otherwise, the message we've received is logged in peer.go.